Update the Applicant model and define a new route

chrislessard opened the pull request:
Hi,

This series of patch introduces a new route for POST-ing and PUT-ing applicants. It's meant to correspond to @dingtaoliu's work on the frontend. One commit also introduces a few new fields to the applicant model. 

Solves #68

freeatnet commented:
Do we really need to write to a file when answering API requests?

freeatnet commented on spec/controllers/export_controller_spec.rb:
```diff
@@ -0,0 +1,163 @@
+        get :transcript_access
+        expect(response.status).to eq(200)
+        expect(response.content_type).to eq("text/csv")
+        expect(response.header["Content-Disposition"]).to eq(
```
Nice!

freeatnet commented on app/controllers/export_controller.rb:
```diff
@@ -4,12 +4,36 @@ class ExportController < ApplicationController
         message: response[:msg]
-      }.to_json
+      }.to_json,
+      content_type: response[:type]
```
Hmm. Is the content type of the `some_hash.to_json` really controlled by `response[:type]`?

michellemtchai replied:
When I didn't add this, the content-type, according the console used to be    unspecified. Is there a more conventional way to set this?

freeatnet replied:
Sorry, what I meant is that the `content_type` is right, but since we're wrapping `response[:msg]` inside of JSON, the `content_type` of the response should always be `application/json`.

freeatnet commented on app/controllers/export_controller.rb:
```diff
@@ -4,12 +4,36 @@ class ExportController < ApplicationController
+  def render_helper(response)
     if response[:generated]
-      send_file("#{Rails.root}/db/seeds/export_data.json")
+      send_data response[:data], :filename => response[:file]
```
We should use `filename: response[:file]` to be consistent with Ruby 1.9+ hash syntax.

michellemtchai replied:
Okay. I'll fix that.

freeatnet commented on spec/services/csv_generator_spec.rb:
```diff
@@ -0,0 +1,237 @@
+        )
+    end
+    context "when there are no assignments in the system" do
+      let (:response) { generator.generate_cdf_info }
```
For next time: we could use the conventional `subject` helper to denote the thing that we're testing. We'd have `subject { something here }` and then `expect(subject).to …`.

michellemtchai replied:
Could you kind of explain how subject helper works?

freeatnet replied:
Replace `let(:response) { generator.generate_cdf_info }` with `subject { generator.generate_cdf_info }`. Now you'll have a `subject` variable that you can test, ie: `expect(subject).to eq({generated: false, …`. It's just a way to explicitly say what the main thing you're testing is.

chrislessard committed:
enhancement(api): Add new fields to applicant model

This patch introduces a new migration adding two new required fields
to the applicant model.

chrislessard committed:
enhancement(api): Define new route for adding applicants

This patch introduces a new route through which applicants can be
POSTed and PUTed.

dingtaoliu reviewed (approved):
Sorry for the late review, the model and controller changes look good to me, and I like that you followed the convention when using context and it in the rspecs!  Just a small nitpick, can you add some dummy values for dept_fields and program into the factory for applicants?  Thanks!

chrislessard committed:
fix(api): Add factories fields for applicant

This commit adds new fields to the factory component for the applicant
model.

dingtaoliu committed:
Merge branch 'master' into TAPP-68
